src/proxy/mod.rs

```diff
@@ -0,0 +1,211 @@
+        "Powered",
+        Value::Boolean(true)
+    ).await?;
+    p.method_call(
+        "org.bluez.Adapter1".try_into()?,
+        "StartDiscovery".try_into()?,
+        []
+    ).await?;
+    Ok(())
+}
+```
+
+*/
+
+mod error;
+
+#[cfg(feature = "introspect")]
+pub mod introspect;
+
+use crate::DBus;
+use dbus_message_parser::message::{Message, MessageType};
+use dbus_message_parser::value::{Bus, Interface, Member, ObjectPath, Value};
+use std::collections::HashMap;
+use std::convert::TryInto;
+
+pub use error::Error;
+pub type ProxyResult<T> = Result<T, Error>;
+
+/// A struct that wraps a connection, destination and path.
+///
+/// A D-Bus "Proxy" is a client-side object that corresponds to a remote object on the server side.
+/// Calling methods on the proxy object calls methods on the remote object.
+/// Read more in the [D-Bus tutorial](https://dbus.freedesktop.org/doc/dbus-tutorial.html#proxies)
+pub struct Proxy<'a> {
+    destination: Bus,
+    object_path: ObjectPath,
+    con: &'a DBus,
+    //timeout
+}
+impl<'a> Proxy<'a> {
+    /// Creates a new proxy struct.
+    pub fn new(destination: Bus, object_path: ObjectPath, con: &'a DBus) -> Self {
+        Proxy {
+            destination,
+            object_path,
+            con,
+        }
+    }
+    pub fn get_object_path(&self) -> &ObjectPath {
+        &self.object_path
+    }
+    pub fn get_destination(&self) -> &Bus {
+        &self.destination
+    }
+    ///Call the RPC method `interface`.`method`(`args`)
+    pub async fn method_call<A>(
+        &self,
+        interface: Interface,
+        method: Member,
+        args: A,
+    ) -> ProxyResult<Message>
+    where
+        A: IntoIterator<Item = Value>,
+    {
+        let mut msg = Message::method_call(
+            self.destination.clone(),
+            self.object_path.clone(),
+            interface,
+            method,
+        );
+        for val in args {
+            msg.add_value(val);
+        }
+        self.call(msg).await
+    }
+    async fn call(&self, msg: Message) -> ProxyResult<Message> {
+        if log::log_enabled!(log::Level::Trace) {
+            if let (Some(path), Some(ifn), Some(mem), Ok(sig)) = (
+                msg.get_path(),
+                msg.get_interface(),
+                msg.get_member(),
+                msg.get_signature(),
+            ) {
+                log::trace!("{}\t{}.{}({:?})", path, ifn, mem, sig);
+            } else {
+                log::trace!("{:?}", msg);
+            }
+        }
+        let resp = self.con.call(msg).await?;
+        //println!("ret: {:?}\n\t({:?})", resp, resp.get_type());
+        log::trace!("ret: {:?} {:?}", resp.get_type(), resp.get_signature());
+        if resp.get_type() == MessageType::Error {
+            return Err(resp.into());
+        }
+        Ok(resp)
+    }
+    ///Get all objects and properties see [D-Bus Specification](https://dbus.freedesktop.org/doc/dbus-specification.html#standard-interfaces-objectmanager)
```

Comment:
Separate the link target like https://github.com/LinkTed/dbus-async/blob/master/src/dbus.rs#L76